crates/node/builder/src/builder/states.rs

```diff
@@ -142,18 +143,58 @@ impl<T: FullNodeTypes, C: NodeComponents<T>> Clone for NodeAdapter<T, C> {
 /// A fully type configured node builder.
 ///
 /// Supports adding additional addons to the node.
-pub struct NodeBuilderWithComponents<T: FullNodeTypes, CB: NodeComponentsBuilder<T>> {
+pub struct NodeBuilderWithComponents<
+    T: FullNodeTypes,
+    CB: NodeComponentsBuilder<T>,
+    AO: NodeAddOns<NodeAdapter<T, CB::Components>>,
+> {
     /// All settings for how the node should be configured.
     pub(crate) config: NodeConfig,
     /// Adapter for the underlying node types and database
     pub(crate) adapter: NodeTypesAdapter<T>,
     /// container for type specific components
     pub(crate) components_builder: CB,
     /// Additional node extensions.
-    pub(crate) add_ons: NodeAddOns<NodeAdapter<T, CB::Components>>,
+    pub(crate) add_ons: AddOns<NodeAdapter<T, CB::Components>, AO>,
```

Comment:
hmm, is there a way to not restrict this at all?
and push the specific type to the `Node` trait?

Reply:
yeah for sure we need it, `RpcContext` in hooks need to know what `EthApi` type its getting, and `EthApi` uses core components